connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+
+import re
+import time
+from contextlib import asynccontextmanager
+from datetime import datetime, timedelta
+from functools import wraps
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+
+from connectors.source import BaseDataSource
+from connectors.utils import CancellableSleeps
+
+BASE_URL = "https://slack.com/api"
+
+CURSOR = "cursor"
+RESPONSE_METADATA = "response_metadata"
+NEXT_CURSOR = "next_cursor"
+DEFAULT_RETRY_SECONDS = 30
+
+LIST_CONVERSATIONS = "list_convos"
+JOIN_CONVERSATIONS = "join_convos"
+CONVERSATION_HISTORY = "convo_history"
+CONVERSATION_REPLIES = "threads"
+LIST_USERS = "list_users"
+
+ENDPOINTS = {
+    LIST_CONVERSATIONS: "conversations.list",
+    JOIN_CONVERSATIONS: "conversations.join",
+    CONVERSATION_HISTORY: "conversations.history",
+    CONVERSATION_REPLIES: "conversations.replies",
+    LIST_USERS: "users.list",
+}
+
+# TODO list
+# Nice to haves:
+# - expand links
+#   - links:read scope needed?
+# - configure which channels to sync via deny list
+# - write an ftest
+
+
+def retryable_aiohttp_call(retries):  # TODO, make a base client class
+    # TODO: improve utils.retryable to allow custom logic
+    # that can help choose what to retry
+    def wrapper(func):
+        @wraps(func)
+        async def wrapped(*args, **kwargs):
+            retry = 1
+            while retry <= retries:
+                try:
+                    async for item in func(*args, **kwargs):
+                        yield item
+                    break
+                except Exception as e:
+                    if retry >= retries:
+                        raise e
+
+                    retry += 1
+
+        return wrapped
+
+    return wrapper
+
+
+class ThrottledError(Exception):
+    """Internal exception class to indicate that request was throttled by the API"""
+
+    pass
+
+
+class SlackClient:
+    def __init__(self, configuration):
+        self.token = configuration["token"]
+        self._http_session = aiohttp.ClientSession(
+            headers=self._headers(),
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._sleeps = CancellableSleeps()
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    def ping(self):
+        return
+
+    def close(self):
+        self._http_session.close()
+        return
+
+    async def list_channels(self, only_my_channels, cursor=None):
+        conversation_batch_limit = 200
+        self._logger.debug("Iterating over all channels")
+        if only_my_channels:
+            self._logger.debug("Will only yield channels the bot belongs to")
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[LIST_CONVERSATIONS]}?limit={conversation_batch_limit}"
```

Comment:
I feel like constants here make it much harder to read, why not inline whole `ENDPOINTS` hash? (`BASE_URL` could be useful still for functional tests)

Reply:
I like using constants for api paths because my chances of typos-that-my-IDE-misses goes way down. Nothing more frustrating than a 404 bug that you track back to a misspelling. 

Would it be more readable if it weren't a dict of endpoints, but just a a bunch of constants, to make it like:
```
f"{BASE_URL}/{LIST_CONVERSATIONS_ENDPOINT}?"
```
?